interp/iostream_test.go

```diff
@@ -0,0 +1,321 @@
+		in := inputFileStream(f)
+		f.Close()
+		if err := in.Close(); err == nil {
+			t.Error("expected inputStream.Close to error")
+		}
+		checkExitCode(t, -1.0, in.ExitCode())
+	})
+}
+
+func TestStreamOutFile(t *testing.T) {
+	t.Run("DoubleClose", func(t *testing.T) {
+		f, err := os.CreateTemp("", "iostream_output*")
+		if err != nil {
+			panic(err)
+		}
+		out := outputFileStream(f, outputStreamBufferSize)
+		f.Close()
+		if err := out.Close(); err == nil {
+			t.Error("expected outputStream.Close to error")
+		}
+		checkExitCode(t, -1.0, out.ExitCode())
+	})
+}
+
+var shellTestCases = []shellTest{
+	randExitCodeTest(),
+	randSignalTest(),
+	shellTest{
+		Name:     "ExitSigPipe",
+		shell:    "exec yes",
+		exitCode: 256 + 13,
+		output:   lines("y"),
+		requires: "yes",
+		skipWin:  true,
+	},
+	shellTest{
+		Name:   "Echo",
+		shell:  "echo loves narcissus",
+		output: lines("loves narcissus"),
+	},
+	shellTest{
+		Name:     "Sort",
+		shell:    "sort -n",
+		input:    "5\n7\n1\n9\n4\n2\n0\n3\n6\n8\n",
+		output:   lines("0", "1", "2", "3", "4", "5", "6", "7", "8", "9"),
+		requires: "sort",
+		skipWin:  true,
+	},
+}
+
+func randExitCodeTest() shellTest {
+	exitCode := rand.Intn(127)
+	shell := fmt.Sprintf("exit %d", exitCode)
+	return shellTest{
+		Name:     "ExitCode",
+		shell:    shell,
+		exitCode: exitCode,
+		output:   emptyOutput(),
+	}
+}
+
+func randSignalTest() shellTest {
+	signal := 1 + rand.Intn(15)
+	shell := fmt.Sprintf("exec kill -%d $$", signal)
+	return shellTest{
+		Name:     "ExitSignal",
+		shell:    shell,
+		exitCode: 256 + signal,
+		output:   emptyOutput(),
+		requires: "kill",
+		skipWin:  true,
+	}
+}
+
+func TestStreamInCmd(t *testing.T) {
+	t.Run("ShellMissing", func(t *testing.T) {
+		fname := randMissingExe()
+		in, err := inputCmdStream(exec.Command(fname))
+		if in != nil {
+			t.Error("should get nil stream on missing command")
+		}
+		checkShellMissing(t, err)
+	})
+
+	for _, tc := range shellTestCases {
+		t.Run(tc.Name, tc.TestInputShell)
+	}
+}
+
+func TestStreamOutCmd(t *testing.T) {
+	t.Run("ShellMissing", func(t *testing.T) {
+		fname := randMissingExe()
+		out, err := outputCmdStream(exec.Command(fname))
+		if out != nil {
+			t.Error("should get nil stream on missing command")
+		}
+		checkShellMissing(t, err)
+	})
+	for _, tc := range shellTestCases {
+		t.Run(tc.Name, tc.TestOutputShell)
+	}
+}
+
+func checkShellMissing(t *testing.T, err error) {
+	if _, ok := err.(*exec.Error); !ok {
+		t.Errorf("error should be *exec.Error, got: %#v", err)
+	}
+	if !strings.Contains(err.Error(), "executable file not found in $PATH") {
+		t.Error("expected error containing \"executable file not found in $PATH\"")
+	}
+}
+
+func checkExitCode(t *testing.T, expectCode, exitCode int) {
+	if exitCode != expectCode {
+		t.Errorf("expected Close() result: %d, got: %d", expectCode, exitCode)
+	}
+}
+
+// Generate a random executable name this is guaranteed not to exist on the $PATH.
+func randMissingExe() string {
```

Comment:
Rather than trying to generate a random executable, can't we just use a fixed name that won't exist? In the existing `TestSystemCommandNotFound`, for example, we just use the name `foobar3982`, which seems highly unlikely to exist. :-)

Reply:
Ok

Reply:
I've removed this test completely.